src/push-manifest.ts

```diff
@@ -0,0 +1,163 @@
+      function shouldIgnoreFile(url: string) {
+        return ignoreUrls && ignoreUrls.has(url);
+      }
+
+      for (const analyzedImport of analyzedImports) {
+        if (!shouldIgnoreFile(analyzedImport.url)) {
+          analyzedImportUrls.add(analyzedImport.url);
+        }
+      }
+      return analyzedImportUrls;
+    }
+
+/**
+ * Gets a link type from a file URL. Useful for generating a PushManifestEntry
+ * and needed for the "as=" value on Link headers.
+ */
+function
+getLinkTypeFromUrl(url: string):
+    LinkType|undefined {
+      return extensionToTypeMapping[path.extname(url)] || undefined;
+    }
+
+/**
+ * Generate a PushManifestEntry from an array of URLs.
+ */
+function
+generatePushManifestEntries(urls: string[]):
+    PushManifestEntry {
+      const newEntry: PushManifestEntry = {};
+      for (const url of urls) {
+        newEntry[url] = {type: getLinkTypeFromUrl(url), weight: 1};
+      }
+      return newEntry;
+    }
+
+/**
+ * Returns a promise that resolves with a generated push manifest, based off of
+ * the options provided.
+ */
+export async function
+generatePushManifest(options: AddPushManifestOptions):
+    Promise<PushManifest> {
+      console.assert(!!options, '`project` & `buildRoot` options are required');
+      console.assert(!!options.project, '`project` option is required');
+      console.assert(!!options.buildRoot, '`buildRoot` option is required');
+
+      const pushManifest: any = {};
+      const analyzer = new Analyzer({
+        urlLoader: new FSUrlLoader(options.buildRoot),
+      });
+
+      // Generate the dependencies to push for the shell
+      const relativeShellUrl = urlFromPath(
+          options.project.config.root, options.project.config.shell);
+      const shellImportUrls =
+          await analyzeFileForImportUrls(analyzer, relativeShellUrl);
+      pushManifest[relativeShellUrl] =
+          generatePushManifestEntries(Array.from(shellImportUrls));
+
+      // Generate the dependencies to push for each fragment.
+      // Make sure that any duplicates from the shell dependencies list are
+      // removed.
+      for (const fragment of options.project.config.fragments) {
+        const relativeFragmentUrl =
+            urlFromPath(options.project.config.root, fragment);
+        const fragmentImportUrls = await analyzeFileForImportUrls(
+            analyzer, relativeFragmentUrl, shellImportUrls);
+        pushManifest[relativeFragmentUrl] =
+            generatePushManifestEntries(Array.from(fragmentImportUrls));
+      }
+
+      return pushManifest;
+    }
+
+/**
+ * Returns a promise that resolves when a push manifest has been generated
+ * and written to the build directory. This uses generatePushManifest() to
+ * generate a push manifest, which it then writes to the file system based on
+ * the buildRoot & path (if provided) options.
+ */
+export async function
+addPushManifest(options: AddPushManifestOptions): Promise<{}> {
```

Comment:
Talked in person, I'm convinced. As an added bonus, the interface is now much easier to use (doesn't need a `buildRoot` because the file stream is all relative).